File: Sources/App/Classes/Dialogs/TDCChannelBanListSheet.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "NSObjectHelperPrivate.h"
#import "TXGlobalModels.h"
#import "TLOLocalization.h"
#import "IRCClient.h"
#import "IRCChannel.h"
#import "IRCISupportInfo.h"
#import "TVCBasicTableView.h"
#import "TDCChannelBanListSheetPrivate.h"

NS_ASSUME_NONNULL_BEGIN

@interface TDCChannelBanListSheetEntry : NSObject
@property(nonatomic, copy) NSString *entryMask;
@property(nonatomic, copy, nullable) NSString *entryMaskDescription; // EXTBAN explanation shown as a tooltip
@property(nonatomic, copy) NSString *entryAuthor;
@property(readonly, copy) NSString *entryCreationDateString;
@property(nonatomic, copy, nullable) NSDate *entryCreationDate;
@end

@interface TDCChannelBanListSheet ()
@property(nonatomic, strong, readwrite) IRCClient *client;
@property(nonatomic, strong, readwrite) IRCChannel *channel;
@property(nonatomic, copy, readwrite) NSString *clientId;
@property(nonatomic, copy, readwrite) NSString *channelId;
@property(nonatomic, assign, readwrite) TDCChannelBanListSheetEntryType entryType;
@property(nonatomic, copy, readwrite, nullable) NSArray<NSString *> *listOfChanges;
@property(nonatomic, weak) IBOutlet NSTextField *headerTitleTextField;
@property(nonatomic, weak) IBOutlet TVCBasicTableView *entryTable;
@property(nonatomic, strong) IBOutlet NSArrayController *entryTableController;

- (IBAction)onUpdate:(nullable id)sender;
- (IBAction)onRemoveEntry:(nullable id)sender;
@end

@implementation TDCChannelBanListSheet

- (nullable instancetype)initWithEntryType:(TDCChannelBanListSheetEntryType)entryType inChannel:(IRCChannel *)channel
{
	NSParameterAssert(channel != nil);

	if ([self.class channel:channel supportsEntryType:entryType] == NO) {
		return nil;
	}

	if ((self = [super initWithWindow:nil])) {
		self.entryType = entryType;

		self.client = channel.associatedClient;
		self.clientId = channel.associatedClient.uniqueIdentifier;

		self.channel = channel;
		self.channelId = channel.uniqueIdentifier;

		[self prepareInitialState];

		return self;
	}

	return nil;
}

- (void)prepareInitialState
{
	[RZMainBundle() loadNibNamed:@"TDCChannelBanListSheet" owner:self topLevelObjects:nil];

	self.entryTable.sortDescriptors = @[ [NSSortDescriptor sortDescriptorWithKey:@"entryCreationDate"
																	   ascending:NO
																		selector:@selector(compare:)] ];

	NSString *headerTitle = nil;

	if (self.entryType == TDCChannelBanListSheetEntryTypeBan) {
		headerTitle = TXTLS(@"TDCChannelBanListSheet[rhc-ke]", self.channel.name);
	} else if (self.entryType == TDCChannelBanListSheetEntryTypeBanException) {
		headerTitle = TXTLS(@"TDCChannelBanListSheet[gbi-wn]", self.channel.name);
	} else if (self.entryType == TDCChannelBanListSheetEntryTypeInviteException) {
		headerTitle = TXTLS(@"TDCChannelBanListSheet[ylc-6e]", self.channel.name);
	} else if (self.entryType == TDCChannelBanListSheetEntryTypeQuiet) {
		headerTitle = TXTLS(@"TDCChannelBanListSheet[g4r-t6]", self.channel.name);
	}

	if (headerTitle == nil) {
		return;
	}

	self.headerTitleTextField.stringValue = headerTitle;
}

- (void)start
{
	[self startSheet];
}

- (void)clear
{
	[self willChangeValueForKey:@"entryCount"];
	[self willChangeValueForKey:@"entryCountDescription"];

	self.entryTableController.content = nil;

	[self didChangeValueForKey:@"entryCountDescription"];
	[self didChangeValueForKey:@"entryCount"];
}

- (void)addEntry:(NSString *)entryMask
		   setBy:(nullable NSString *)entryAuthor
	creationDate:(nullable NSDate *)entryCreationDate
{
	NSParameterAssert(entryMask != nil);

	if (entryAuthor == nil) {
		entryAuthor = TXTLS(@"BasicLanguage[vbl-xi]"); // "Unknown"
	}

	TDCChannelBanListSheetEntry *newEntry = [TDCChannelBanListSheetEntry new];

	newEntry.entryMask = entryMask;
	newEntry.entryMaskDescription = [self.client.supportInfo descriptionForExtendedBanMask:entryMask];
	newEntry.entryAuthor = entryAuthor;
	newEntry.entryCreationDate = entryCreationDate;

	[self willChangeValueForKey:@"entryCount"];
	[self willChangeValueForKey:@"entryCountDescription"];

	[self.entryTableController addObject:newEntry];

	[self didChangeValueForKey:@"entryCountDescription"];
	[self didChangeValueForKey:@"entryCount"];
}

- (NSNumber *)entryCount
{
	return @([self.entryTableController.arrangedObjects count]);
}

/* "N entries", or "N of MAX entries" when the server advertises MAXLIST. */
- (NSString *)entryCountDescription
{
	NSUInteger entryCount = [self.entryTableController.arrangedObjects count];

	NSUInteger maximumEntries = [self.client.supportInfo maximumListEntriesForModeSymbol:self.modeSymbol];

	if (maximumEntries > 0) {
		return TXTLS(
			@"TDCChannelBanListSheet[n0f-mx]", TXFormattedNumber(entryCount), TXFormattedNumber(maximumEntries));
	}

	return TXTLS(@"TDCChannelBanListSheet[n0f-cn]", TXFormattedNumber(entryCount));
}

#pragma mark -
#pragma mark Actions

- (void)onUpdate:(nullable id)sender
{
	[self clear];

	if ([self.delegate respondsToSelector:@selector(channelBanListSheetOnUpdate:)]) {
		[self.delegate channelBanListSheetOnUpdate:self];
	}
}

- (void)onRemoveEntry:(nullable id)sender
{
	NSIndexSet *selectedRows = self.entryTable.selectedRowIndexes;

	NSMutableArray<NSString *> *selectedEntries = [NSMutableArray arrayWithCapacity:selectedRows.count];

	[selectedRows enumerateIndexesUsingBlock:^(NSUInteger index, BOOL *stop) {
		TDCChannelBanListSheetEntry *entryItem = self.entryTableController.arrangedObjects[index];

		[selectedEntries addObject:entryItem.entryMask];
	}];

	self.listOfChanges = [self.client compileListOfModeChangesForModeSymbol:self.modeSymbol
																  modeIsSet:NO
															 modeParameters:selectedEntries];

	[super cancel:nil];
}

#pragma mark -
#pragma mark Utilities

+ (BOOL)channel:(IRCChannel *)channel supportsEntryType:(TDCChannelBanListSheetEntryType)entryType
{
	return [channel.associatedClient.supportInfo isListSupported:(IRCISupportInfoListType)entryType];
}

- (NSString *)modeSymbol
{
	/* -modeSymbolForList: is nullable but because we only allow this class to be
	 created if the mode is already supported, then we can advertise it here as non-nil */
	return [self.client.supportInfo modeSymbolForList:(IRCISupportInfoListType)self.entryType];
}

#pragma mark -
#pragma mark NSWindow Delegate

- (void)windowWillClose:(NSNotification *)note
{
	if ([self.delegate respondsToSelector:@selector(channelBanListSheetWillClose:)]) {
		[self.delegate channelBanListSheetWillClose:self];
	}
}

@end

#pragma mark -

@implementation TDCChannelBanListSheetEntry

- (NSString *)entryCreationDateString
{
	NSDate *entryCreationDate = self.entryCreationDate;

	if (entryCreationDate == nil) {
		return TXTLS(@"BasicLanguage[vbl-xi]"); // "Unknown"
	}

	return TXFormatDateLongStyle(entryCreationDate, YES);
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/AppKitSupport/TDCChannelBanListSheet.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import AppKit

/// What `ChannelBanListSheet` reports back.
@MainActor
public protocol ChannelBanListSheetDelegate: AnyObject {
	func channelBanListSheetOnUpdate(_ sender: ChannelBanListSheet)
	func channelBanListSheetWillClose(_ sender: ChannelBanListSheet)
}

@objc(TDCChannelBanListSheetEntryType)
public enum ChannelBanListEntryType: UInt {
	case ban = 0
	case banException
	case inviteException
	case quiet

	/** Spelled out rather than mapped by raw value: the two enums are declared
	 in different modules and reordering either one would silently point the
	 sheet at the wrong mode. */
	var supportListType: IRCISupportInfoListType {
		switch self {
		case .ban: .ban
		case .banException: .banException
		case .inviteException: .inviteException
		case .quiet: .quiet
		}
	}
}

@objc(TDCChannelBanListSheetEntry)
public final class ChannelBanListSheetEntry: NSObject {
	@objc public var entryMask = ""
	@objc public var entryMaskDescription: String?
	@objc public var entryAuthor = ""
	@objc public var entryCreationDate: Date?

	@objc public var entryCreationDateString: String {
		guard let entryCreationDate else {
			return ApplicationStrings.unknownValue
		}

		return formatDateLongStyle(entryCreationDate, true) ?? ApplicationStrings.unknownValue
	}
}

@objc(TDCChannelBanListSheet)
@MainActor
public final class ChannelBanListSheet: SheetBase, TDCChannelPrototype {
	@objc public private(set) var client: IRCClient!
	@objc public private(set) var channel: IRCChannel!
	@objc public private(set) var clientId: String?
	@objc public private(set) var channelId: String?
	@objc public private(set) var entryType: ChannelBanListEntryType = .ban
	@objc public private(set) var listOfChanges: [String]?
	@objc public var contentAlreadyReceived = false

	@IBOutlet private var headerTitleTextField: NSTextField!
	@IBOutlet private var entryTable: BasicTableView!
	@IBOutlet private var entryTableController: NSArrayController!

	private var tableEntries: [Any] {
		entryTableController.arrangedObjects as? [Any] ?? []
	}

	@objc(initWithEntryType:inChannel:)
	public init?(entryType: ChannelBanListEntryType, inChannel channel: IRCChannel) {
		guard Self.channel(channel, supportsEntryType: entryType) else {
			return nil
		}

		super.init(window: nil)

		self.entryType = entryType
		client = channel.associatedClient
		clientId = channel.associatedClient!.uniqueIdentifier
		self.channel = channel
		channelId = channel.uniqueIdentifier

		prepareInitialState()
	}

	private func prepareInitialState() {
		Bundle.main.loadNibNamed("TDCChannelBanListSheet", owner: self, topLevelObjects: nil)

		entryTable.sortDescriptors = [
			NSSortDescriptor(key: "entryCreationDate", ascending: false, selector: #selector(NSDate.compare(_:))),
		]

		headerTitleTextField.stringValue = ChannelAccessListStrings.heading(
			for: entryType,
			channelName: channel.name
		)
	}

	@objc public func start() {
		startSheet()
	}

	@objc public func clear() {
		willChangeValue(forKey: "entryCount")
		willChangeValue(forKey: "entryCountDescription")

		entryTableController.content = nil

		didChangeValue(forKey: "entryCountDescription")
		didChangeValue(forKey: "entryCount")
	}

	@objc(addEntry:setBy:creationDate:)
	public func addEntry(
		_ entryMask: String,
		setBy entryAuthor: String?,
		creationDate entryCreationDate: Date?
	) {
		var author = entryAuthor

		if author == nil {
			author = ApplicationStrings.unknownValue
		}

		let newEntry = ChannelBanListSheetEntry()
		newEntry.entryMask = entryMask
		newEntry.entryMaskDescription = client.supportInfo.descriptionForExtendedBanMask(entryMask)
		newEntry.entryAuthor = author!
		newEntry.entryCreationDate = entryCreationDate

		willChangeValue(forKey: "entryCount")
		willChangeValue(forKey: "entryCountDescription")

		entryTableController.addObject(newEntry)

		didChangeValue(forKey: "entryCountDescription")
		didChangeValue(forKey: "entryCount")
	}

	@objc public dynamic var entryCount: NSNumber {
		NSNumber(value: tableEntries.count)
	}

	@objc public dynamic var entryCountDescription: String {
		let entryCount = tableEntries.count
		let maximumEntries = Int(client.supportInfo.maximumListEntries(forModeSymbol: modeSymbol))

		return ChannelAccessListStrings.entryCount(entryCount, maximum: maximumEntries)
	}

	@IBAction private func onUpdate(_: Any?) {
		clear()

		banListDelegate?.channelBanListSheetOnUpdate(self)
	}

	@IBAction private func onRemoveEntry(_: Any?) {
		let selectedRows = entryTable.selectedRowIndexes
		var selectedEntries: [String] = []

		for index in selectedRows {
			guard let entryItem = tableEntries[index] as? ChannelBanListSheetEntry else {
				continue
			}

			selectedEntries.append(entryItem.entryMask)
		}

		listOfChanges = client.compileListOfModeChanges(
			forModeSymbol: modeSymbol,
			modeIsSet: false,
			modeParameters: selectedEntries
		)

		super.cancel(nil)
	}

	private var banListDelegate: (any ChannelBanListSheetDelegate)? {
		delegate as? any ChannelBanListSheetDelegate
	}

	@objc(channel:supportsEntryType:)
	public static func channel(
		_ channel: IRCChannel,
		supportsEntryType entryType: ChannelBanListEntryType
	) -> Bool {
		guard let client = channel.associatedClient else {
			return false
		}

		return client.supportInfo.isListSupported(entryType.supportListType)
	}

	@objc public var modeSymbol: String {
		client.supportInfo.modeSymbol(forList: entryType.supportListType) ?? ""
	}

	@objc public func windowWillClose(_: Notification) {
		banListDelegate?.channelBanListSheetWillClose(self)
	}
}
